### ISSUE TITLE
Improve helm chart failure cases

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#4989]
We should improve how we handle and log failure conditions for charts. Also it points to bad caching possibly.

One repo triggers this error 20 times in logs:
```json
    "err": {
        "name": "RequestError",
        "code": "EAI_AGAIN",
        "hostname": "charts.cloud.company",
        "method": "GET",
        "path": "/index.yaml",
        "protocol": "https:",
        "url": "https://charts.cloud.company/index.yaml",
        "gotOptions": {
            "path": "/index.yaml",
            "protocol": "https:",
            "hostname": "charts.cloud.company",
            "hash": "",
            "search": "",
            "pathname": "/index.yaml",
            "href": "https://charts.cloud.company/index.yaml",
            "headers": {
                "user-agent": "Renovate Bot (GitHub App 2740)",
                "accept-encoding": "gzip, deflate"
            },
            "hooks": {
                "beforeError": [],
                "init": [],
                "beforeRequest": [],
                "beforeRedirect": [],
                "beforeRetry": [],
                "afterResponse": []
            },
            "retry": {
                "methods": {},
                "statusCodes": {},
                "errorCodes": {}
            },
            "decompress": true,
            "throwHttpErrors": true,
            "followRedirect": true,
            "stream": false,
            "form": false,
            "json": false,
            "cache": false,
            "useElectronNet": false,
            "baseUrl": "https://charts.cloud.company/",
            "method": "GET"
        },
        "message": "getaddrinfo EAI_AGAIN charts.cloud.company charts.cloud.company:443",
        "stack": "RequestError: getaddrinfo EAI_AGAIN charts.cloud.company charts.cloud.company:443\n    at ClientRequest.request.once.error (/home/ubuntu/renovateapp/node_modules/got/source/request-as-event-emitter.js:178:14)\n    at Object.onceWrapper (events.js:286:20)\n    at ClientRequest.emit (events.js:203:15)\n    at ClientRequest.EventEmitter.emit (domain.js:448:20)\n    at ClientRequest.origin.emit.args (/home/ubuntu/renovateapp/node_modules/@szmarczak/http-timer/source/index.js:37:11)\n    at TLSSocket.socketErrorListener (_http_client.js:392:9)\n    at TLSSocket.emit (events.js:198:13)\n    at TLSSocket.EventEmitter.emit (domain.js:448:20)\n    at emitErrorNT (internal/streams/destroy.js:91:8)\n    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)"
    },
    "msg": "https://charts.cloud.company lookup failure: Unknown error",
```

First, it's not really an Unknown error - the host couldn't be resolved.
Second, it seems like we query the same URL over and over.

**Comment by rarkins** [comment]
@nchashch are you still available to look at this?

**Comment by rarkins** [comment]
The same repo triggers this additional warning: `"msg": "Couldn't get index.yaml file from https://charts.cloud.company",`. Ideally we could have "standardized" warning messages across datasources so it's easier to filter them all.

**Comment by nchashch** [comment]
@rarkins I will look into this tomorrow.

**Comment by rarkins** [comment]
I’ve made some changes already to lower the warnings. Need to check if we are caching enough though

**Comment by nchashch** [comment]
TL;DR: `index.yaml` wasn't taken from cache because all requests to https://charts.cloud.company/index.yaml failed, and `index.yaml` was never written to cache.

@rarkins Caching in helm datasource works like in other datasources:
```ts
  const cacheNamespace = 'datasource-helm';
  const cacheKey = repository;
  const cachedIndex = await renovateCache.get(cacheNamespace, cacheKey);
  if (cachedIndex) {
    return cachedIndex;
  } else {
  // Attempt fetching and parsing index.yaml
  // If everything is correct:
    const cacheMinutes = 20;
    await renovateCache.set(cacheNamespace, cacheKey, result, cacheMinutes);
    return result;
  }
```
The difference is that `index.yaml` contains data for all charts in a `repository`, not a single package, like in other datasources. But cache is accessed by `repository`, which is the same for all charts from one repository, so it should only make one request per repository.

I tried openening https://charts.cloud.company/index.yaml in the browser, and got `DNS_PROBE_FINISHED_NXDOMAIN`. It looks like an invalid link.

So if 20 charts from `charts.cloud.company` are processed by `renovate` and requests for https://charts.cloud.company/index.yaml fail every time, `index.yaml` will never be written to the cache and so it will keep making requests, until at least one succeeds.

If first request succeeded, then `index.yaml`, would have been written to the cache, and for the next 20 minutes all calls to `getPkgReleases` would have used the cached version.

**Comment by HonkingGoose** [comment]
Are we still having issues with the current Renovate code? Last comment was from December 2019, so might be worth taking a look again?
